Add message rfc (#17)
iotaledger/protocol-rfcs: text/0017-message/0017-message.md

+ Feature name: `tangle-message`
+ Start date: 2020-07-28
+ RFC PR: [iotaledger/protocol-rfcs#0017](https://github.com/iotaledger/protocol-rfcs/pull/0017)

# Summary

> A message is the object nodes gossip around in the network. It always references two other messages that are known as _parents_. It is stored as a vertex on the tangle data structure that the nodes maintain.

Reviewer: > It always references two other messages that are known as _parents_.

Not factually correct

+ Feature name: `tangle-message`
+ Start date: 2020-07-28
+ RFC PR: [iotaledger/protocol-rfcs#0017](https://github.com/iotaledger/protocol-rfcs/pull/0017)

# Summary

A message is the object nodes gossip around in the network. It always references two other messages that are known as `parents`. It is stored as a vertex on the tangle data structure that the nodes maintain.

> The messages will contain payloads. Some of them will be core payloads that will be processed by all nodes as part of the core protocol. Some of them will be community payloads that will enable to build new functionality on top of the tangle. Some payloads may have other nested payloads embedded inside.

Reviewer 2: ```suggestion
The messages will contain payloads. Some of them will be core payloads that will be processed by all nodes as part of the core protocol. Some of them will be community payloads that will enable to build new functionality on top of the Tangle. Some payloads may have other nested payloads embedded inside.
```

| --------         | -----------   | -----------               |
| Payload Type     | uint32        | Must be set to **2**      |
| Index            | string     | The index key of the message |
| Data             | ByteArray     | Data we are attaching    |

Note that `index` must be at least 1 byte long for the payload to be valid. The `data` may have a length of 0.

### Serialization Example

Below is a serialized valid message with the indexation payload. The index is the "SPAM" ASCII string and the message is the "Hello Iota"
ASCII string. The [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) would have
  10 trailing zeroes for the given nonce in this example. Bytes are expressed as hexadecimal numbers.

[Version] **`01`** [Parent 1] `F532A53545103276B46876C473846D98648EE418468BCE76DF4868648DD73E5D` [Parent 2] `78D546B46AEC4557872139A48F66BC567687E8413578A14323548732358914A2` [Payload Length]
*`1100000000000000000000000000000000000000000000000000000000000000`*[Payload Type]**`02`**[Index] *`04`*`5350414D` [Data]*`0A`*`48656C6C6F20496F7461`[Nonce]`5d38333333333333`

# Rationale and alternatives

> Instead of creating a layered approach, we could have simply created a flat `transaction message`, that is tailored for mutating the ledger state, and try to fit all the use cases there. For example with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.

Reviewer 2: ```suggestion
Instead of creating a layered approach, we could have simply created a flat `transaction message` that is tailored for mutating the ledger state, and try to fit all the use cases there. For example, with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction, we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.
```

| uint32  | An unsigned 32 bit integer encoded in Little Endian. |
| uint64  | An unsigned 64 bit integer encoded in Little Endian. |
| ByteArray[N] | A static size array of size N.   |
| ByteArray | A dynamically sized array. A uint32 denotes its length.   |
| string | A dynamically sized array of an UTF-8 encoded string. A uint16 denotes its length.   |

### Message ID
The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.

### Message structure

<table>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Version</td>
        <td>uint8</td>
        <td>The message version. The schema specified in this RFC is for version <strong>1</strong> only. </td>
    </tr>
    <tr>
        <td>Parent1 (<code>trunk</code>)</td>
        <td>ByteArray[32]</td>
        <td>The Message ID of the first <i>Message</i> it references.</td>
    </tr>
    <tr>
        <td>Parent2 (<code>branch</code>)</td>
        <td>ByteArray[32]</td>
        <td>The Message ID of the second <i>Message</i> it references.</td>
    </tr>
    <tr>

>         <td>Payload Length</td>

Reviewer 3: is it memory length or serialization length of the payload?

Author: serialization length

After deserializing the payload, in theory a node implementation can choose any way it wants to represent it in memory.
We attempt to spec the protocol, not the implementation details (unless they are necessary)

3. If the `payload type` is in the core payload range (0-127) and the node is familiar with it, or if it is above this range.
4. If the [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) will contain at least the number of trailing 0 trits the node defines as required.

### Payloads

A message may contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `indexation payload` will be specified here as an example.

| Payload Name                              |   Type Value |
| ---------------------------------------   | -----------  | 
|  [Signed Transaction](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-signed-transaction-payload/0000-signed-transaction-payload.md)                       |     0        |
|  [Milestone Draft](https://github.com/jakubcech/protocol-rfcs/blob/jakubcech-milestonepayload/text/0019-milestone-payload/0019-milestone-payload.md)                                |     1        |

> |  [Indexation Payload](#Unsigned-Data-Payload)  |     2        |

Reviewer 3: I think it refer to indexation payload.
```suggestion
|  [Indexation Payload](#indexation-payload)  |     2        |
```

+ Feature name: `tangle-message`
+ Start date: 2020-07-28
+ RFC PR: [iotaledger/protocol-rfcs#0017](https://github.com/iotaledger/protocol-rfcs/pull/0017)

# Summary

A message is the object nodes gossip around in the network. It always references two other messages that are known as `parents`. It is stored as a vertex on the tangle data structure that the nodes maintain.

> The messages will contain payloads. Some of them will be core payloads that will be processed by all nodes as part of the core protocol. Some of them will be community payloads that will enable the building of new functionalities on top of the Tangle. Some payloads may have other nested payloads embedded inside.

Reviewer 4: In the Payloads section below is says "A message _may_ contain a payload" but here it says "will", could you clarify if it is strictly required or not?

| uint16  | An unsigned 16 bit integer encoded in Little Endian. |
| uint32  | An unsigned 32 bit integer encoded in Little Endian. |
| uint64  | An unsigned 64 bit integer encoded in Little Endian. |
| ByteArray[N] | A static size array of size N.   |
| ByteArray | A dynamically sized array. A uint32 denotes its length.   |
| string | A dynamically sized array of an UTF-8 encoded string. A uint16 denotes its length.   |

### Message ID
The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.

### Message structure

<table>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>NetworkID</td>
        <td>uint64</td>
        <td>Network identifier. This field will signify whether this message was meant for mainnet, testnet, or a private net. It also tells what protocol rules apply to the message. It is first 8 bytes of the `BLAKE2b-256` hash of the concatenation of the network type and the protocol version string.</td>
        </tr>
    <tr>
        <td> Parents' length </td>
        <td> uint8</td>

>         <td> The number of messages we directly approve. Can be any value between 1-8.</td>

Reviewer 4: Could you clarify the exact range here a little bit? It's not immediately obvious if "between 1-8" is inclusive or exclusive (or half open), and in the first sentence of the summary it says "It always references two other messages that are known as `parents`".

Maybe something like: "Can be any value in the range (1,8]"?

    </tr>
    <tr>
        <td>Parents </td>
        <td>ByteArray[32 * `parents length`]</td>

>         <td>The Message IDs that are referenced.</td>

Reviewer 4: Could the description be expanded to say that these Message IDs are the ones directly approved? Currently that information is only in the "Parents' length" spec above.

Reviewer 5: done

                        <td>
                            The type of the payload. It will instruct the node how to parse the fields that follow.
                        </td>
                    </tr>
                    <tr>
                        <td>Data Fields</td>
                        <td>ANY</td>
                        <td>A sequence of fields, where the structure depends on <code>payload type</code>.</td>
                    </tr>
                </table>
            </details>
    <tr>
        <td>Nonce</td>
        <td>uint64</td>
        <td>The nonce which lets this message fulfill the Proof-of-Work requirement.</td>
    </tr>
</table>

### Message validation

A message is considered valid, if the following syntactic rules are met:

1. The message size must not exceed 32 KiB (32 * 1024 bytes).
2. When parsing the message is complete, there should not be any trailing bytes left that were not parsed.
3. If the `payload type` is known to the node.
4. If the [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) will contain at least the number of trailing 0 trits the node defines as required.

> 5. `Parents' length` must be between 1-8.

Reviewer 4: Same comment here re. clarification of allowed range.

### Payloads

> A message may contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `indexation payload` will be specified here as an example:

Reviewer 4: Here it say it "may" contain a payload, what would the message format be if no payload exists? It seems like by default the rest of the message would be zero'd out which could be interpreted as a Signed Transaction (according to the table below).

| Payload Name                              |   Type Value |
| ---------------------------------------   | -----------  | 
|  [Signed Transaction](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-signed-transaction-payload/0000-signed-transaction-payload.md)                       |     0        |
|  [Milestone Draft](https://github.com/jakubcech/protocol-rfcs/blob/jakubcech-milestonepayload/text/0019-milestone-payload/0019-milestone-payload.md)                                |     1        |

> |  [Indexation Payload](#indexation-payload)  |     2        |

Reviewer: Should we add the 2 new payloads ?

To be clear, the concept of the Payload allows the addition of an index to the encapsulating message, as well as some arbitrary data. Nodes will expose an API that will enable the querying of messages by the index.
Adding those capabilities may open nodes to DOS attack vectors:
1. Proliferation of index keys that may blow up the node's DB
2. Proliferation of messages associated with the same index

Node implementations may provide weak guarantees regarding the completion of indexes to address the above scenarios. 

Besides the index, the payload will also have a data field.
  A message that has been attached to the Tangle and approved by a milestone has several useful properties: verifying that the content of the data did not change and determining the approximate time it was published by checking the approving milestone. If the payload will be incorporated under
  the `signed transaction payload`, the content will be signed as well.

The structure of the payload is simple:

| Name             | Type          | Description               |
| --------         | -----------   | -----------               |
| Payload Type     | uint32        | Must be set to **2**      |

> | Index            | ByteArray     | The index key of the message |

Reviewer: `ByteArray` has a length denoted by a `u32` but index should be denoted by a `u16`.

  10 trailing zeroes for the given nonce in this example. Bytes are expressed as hexadecimal numbers.

[Version] **`01`** [Parent 1] `F532A53545103276B46876C473846D98648EE418468BCE76DF4868648DD73E5D` [Parent 2] `78D546B46AEC4557872139A48F66BC567687E8413578A14323548732358914A2` [Payload Length]
*`1100000000000000000000000000000000000000000000000000000000000000`*[Payload Type]**`02`**[Index] *`04`*`5350414D` [Data]*`0A`*`48656C6C6F20496F7461`[Nonce]`5d38333333333333`

# Rationale and alternatives

Instead of creating a layered approach, we could have simply created a flat `transaction message` that is tailored for mutating the ledger state, and try to fit all the use cases there. For example, with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction, we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.

This approach seems less extensible. It might have made sense if we wanted to build a protocol that is just for ledger mutating transactions, but we want to be able to extend the protocol to do more than that.

# Unresolved questions

> - What should be the maximum length of the message?

Reviewer 4: Is this resolved? In the text above it says "The message size must not exceed 32 KiB (32 * 1024 bytes)"

  the `signed transaction payload`, the content will be signed as well.

The structure of the payload is simple:

| Name             | Type          | Description               |
| --------         | -----------   | -----------               |
| Payload Type     | uint32        | Must be set to **2**      |
| Index            | string     | The index key of the message |
| Data             | ByteArray     | Data we are attaching    |

Note that `index` and `data` may both have a length of 0.
There are no validation rules for the payload. Message validation rules suffice here.

### Serialization Example

Below is a serialized valid message with the indexation payload. The index is the "SPAM" ASCII string and the message is the "Hello Iota"
ASCII string. The [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) should have
  14 trailing zeroes. Bytes are expressed as hexadecimal numbers. 

[Version] **`01`** [Parent 1] `F532A53545103276B46876C473846D98648EE418468BCE76DF4868648DD73E5D` [Parent 2] `78D546B46AEC4557872139A48F66BC567687E8413578A14323548732358914A2` [Payload Length]

> *`0C`*[Payload Type]**`02`**[Index] *`04`*`9A52` [Data]*`0A`*`48656c6c6f20496f7461`[Nonce]`dc293f3333333333`

Reviewer: `9A52` == `SPAM` ?

Author: Hmm...
Not sure how this happened :-|

Also the payload length is incorrect...
I will change the amount of trailing 0s to 10 for the example because it is easier to calculate

Reviewer: Also, the index length and data length seem to be only one byte ?

Author: Before I fix, I am starting to wonder whether the 32 bytes we specified for the length is a good decision...
I understand it is convenient, but maybe it is just bloating up message size for no good reason?

> + Feature name: `Tangle Messages`

Reviewer: ```suggestion
+ Feature name: `tangle-message`
```
For consistency

+ Start date: 2020-07-28
+ RFC PR: [iotaledger/protocol-rfcs#0017](https://github.com/iotaledger/protocol-rfcs/pull/0017)

# Summary

A message is the object nodes gossip around in the network. It always references two other messages that are known as `parents`. It is stored as a vertex on the tangle data structure that the nodes maintain.

The messages will contain payloads. Some of them will be core payloads that will be processed by all nodes as part of the core protocol. Some of them will be community payloads that will enable to build new functionality on top of the tangle.

Some payloads may have other nested payloads embedded inside.

> So upon parsing it is done layer by layer.

Reviewer: Merge this sentence with the previous one.

| uint64  | An unsigned 64 bit integer encoded in Little Endian.
| ByteArray    | A dynamic size byte array. It is a prefixed by a `varint` that indicates the number of bytes in the array. The following bytes are simply the content of the array.
| ByteArray[N] | A static size array of size N. Since the size is expected, no need to prefix with a varint   |

### Message ID
The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.

### Message Structure

<table>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Version</td>
        <td>varint</td>
        <td>The message version. The schema specified in this RFC is for version <strong>1</strong> only. </td>
    </tr>
    <tr>
        <td>Parent1 (<code>trunk</code>)</td>
        <td>ByteArray[32]</td>

>         <td>The Message ID of the first <i>Message</i> we reference.</td>

Reviewer: ```suggestion
        <td>The Message ID of the first <i>Message</i> it references.</td>
```

    </tr>
    <tr>
        <td>Parent2 (<code>branch</code>)</td>
        <td>ByteArray[32]</td>

>         <td>The Message ID of the second <i>Message</i> we reference.</td>

Reviewer: ```suggestion
        <td>The Message ID of the second <i>Message</i> it references.</td>
```

                        <th>Description</th>
                    </tr>
                    <tr>
                        <td>Payload Type</td>
                        <td>varint</td>
                        <td>
                            The type of the payload. It will instruct the node how to parse the fields that follow. Types in the range of 0-127 are "core types" that all nodes are expected to know.
                        </td>
                    </tr>
                    <tr>
                        <td>Data Fields</td>
                        <td>ANY</td>
                        <td>A sequence of fields, where the structure depends on <code>payload type</code>.</td>
                    </tr>
                    <tr>
                        <td>Nested Payload (Optional)</td>
                        <td> Payload</td>
                        <td> Some payloads support nesting other payloads inside of them.</td>
                    </tr>
                </table>
            </details>
    <tr>
        <td>Nonce</td>
        <td>uint64</td>
        <td>The nonce which lets this message fulfill the Proof-of-Work requirement.</td>
    </tr>
</table>

> ### Message PoW Hash

Reviewer 5: The message PoW is now in its own RFC #24.
I think it would be best, to just remove this section. I wouldn't even reference #24 since this introduces a circular dependency and the exact PoW computation is not relevant in the context of this RFC.

> This hash is computed by converting all the messages raw bytes into trytes with `b1t6`. Then hash the result with Curl-P-81. The result will be used to verify the proof-of-work done on the message. This will be used for spam protection.

Reviewer 6: What is the reasoning behind continuing to use Curl-P-81?

Author: Changing to a different PoW may open us up to spam attacks from people who possess specialized hardware.
With Curl-P-81 we know exactly what we are dealing with at the very least

We may switch from it later on after we have a conclusion to what PoW algo we want to switch to

Reviewer 5: This is also motivated in #24

> This hash is computed by converting all the messages raw bytes into trytes with `b1t6`. Then hash the result with Curl-P-81. The result will be used to verify the proof-of-work done on the message. This will be used for spam protection.

Reviewer: We usually put the `t` before the `b`, so `t6b1`.

Reviewer: Also, trits, not trytes.

Reviewer 5: I think this is fundamentally different:
- `t5b1` encodes ternary data into binary, by converting each group of 5 trits into one byte; it cannot decode arbitrary binary data
- `b1t6` encodes binary data into ternary, by converting each byte into 6 trits; it cannot decode arbitrary ternary data

In [RFC-15](https://iotaledger.github.io/protocol-rfcs/0015-binary-to-ternary-encoding/0015-binary-to-ternary-encoding.html) it is also called `b1t6`.

Author: I think maybe I can delete this entire section. Now we have an RFC dedicated for computing this PoW hash

Reviewer 5: Yes, the entire section should be deleted. See https://github.com/iotaledger/protocol-rfcs/pull/17#discussion_r476267452

Author: due to me looking at the RFC while you are commenting, caused me to repeat your remarks

### Message Validation

A message is considered valid, if the following syntactic rules are met:

1. The message length must not exceed X [tbd] bytes.
2. Varint fields must adhere to the following rules:
    * *Malleability protection:* No `0x80` or `0x00` trailing bytes. Take for example the following 3 byte varint: `0x8F8000`. 
    In order to parse it we look at its bits: 
    `1000 1111 1000 0000 0000 0000`.
    First we strip the msbs:
    `000 1111 000 0000 000 0000`
    Then we reverse the groups of 7 bits (Little Endian):
    `000 0000 000 0000 000 1111`
    Then we concatenate all the bits to finally get:
    `1111`
    So the varint `0x0F` equals to `0x8F8000`, but is encoded with different bytes.
    * *Size Protection*: The varint must not be more than 10 bytes long. A 10 byte varint can encompass all the range of `uint64`. In case of an overflow above the `uint64` range the message is invalid.
2. When we are done parsing the message there shouldn't be any trailing bytes left that were not parsed.

> 4. If the `payload type` is in the core payload range (0-127) and the node is familiar with it, or if it is in the range of 128-16383. The upper limit of 16383 is the maximal value of a 2 byte varint, and it should suffice for all the optional payload types we need.

Reviewer 6: How do register a custom (community) payload type?

Is 16383 sufficient for all applications?

Author: IIRC, in our meetings we threw ideas on how it might be.
I think one idea was that community member will also create RFCs but not sure this is the best way to approach this

I think that together with community volunteers we should define at least one community payload. Then we should properly document and implement it, so other community members can easily follow.
We should probably talk about it more in a meeting.

Regarding the 16,383 (2^14 - 1) limit, I just thought it was a high enough number...
Given that we support colored coins it is hard for me to imagine that the payload feature will become so popular that this limit won't suffice...

Reviewer 5: We should really explain how we envision this community-driven payload draft process.
I guess we want to have a table somewhere where people can create a PR with a new ID and link to the document / RFC describing the payload.

Does it maybe make more sense to have a separate RFC with this table (similar to [SLIP-44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) or [SLIP-173](https://github.com/satoshilabs/slips/blob/master/slip-0173.md))? That could make the process considerably easier.

Reviewer 7: Why use a varint if you have a max bound of 2 bytes? It is way easier to define a uint16 here

Author: See my other answer

Also Wolfgang seems to not like this bound, so maybe this bound will be removed (still need to discuss this with him)

Reviewer 5: I think the bound (for the non-core payloads) seems arbitrary and should be removed. We should allow for the full uint32 maybe even uint64 space.

Author: I agree the message length protection suffices.
Actually, while thinking about the case for the bound, I started to question whether the message length is really needed...

5. If the Message PoW Hash will contain at least the number of trailing 0 trits the node defines as required.

### Payloads

> Each message will contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `Unsigned Payload` will be specified here as an example.

Reviewer: ```suggestion
Each message will contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `Unsigned Data Payload` will be specified here as an example.
```

> Each message will contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `Unsigned Payload` will be specified here as an example.

Reviewer 5: ```
Each message will contain a payload.
```
Messages without payload data should be permitted. They will just have a payload length of zero but are otherwise 100% compatible to the specification.

Each message will contain a payload. The specification of the payloads is out of scope of this RFC. Below is a table of the currently specified core payloads with a link to their specifications. The `Unsigned Payload` will be specified here as an example.

| Payload Name                              |   Type Value |
| ---------------------------------------   | -----------  | 
|  Signed Transaction                       |     0        |
|  Signed Data                              |     1        |
|  [Unsigned Data](#Unsigned-Data-Payload)  |     2        |

> | Indexation                                |     3        |

Author: Type value can still be changed
This is just temporary... when I left it blank people asked kept on asking me why it is blank

I expect that in a meeting we will sort out the values

Reviewer: Is the Milestone payload missing though ?

### Unsigned Data Payload

To make the Payload concept clear we will define the `unsigned data payload`. As the name suggests it simply allows to add arbitrary data to a message. A message that has been attached to the tangle and approved by a milestone has useful properties: You can verify that the content of the data did not change, and you can ascertain the approximate time it was published by checking the approving milestone. For example, one can record a hash of a text document on the tangle with such a payload. When a third party requests a proof that the document was published at a certain time and was not changed, one can point them to them to the message containing the `unsigned data payload` with the hash.

The structure of the payload is simple:

> 

Reviewer: Why 4 new lines :) ?

| Name             | Type     | Description               |
| --------         | -------- | -----------               |
| Payload Type     | varint   | Must be set to **2**      |
| Data             | ByteArray| The data we are attaching |

> There are no validation rules for the payload. Message validation rules suffice here.

Reviewer 8: Does this mean that an `Unsigned Data Payload` with a byte array of length zero is valid?

Author: If the `payload length` field in the message accommodates only for the type then in my opinion it should be valid.

Obviously there is no point in creating such a payload, but I don't see an issue here

> 

Reviewer: Why 4 new lines :) ?

Author: I think it renders the same, but I will delete

# Rationale and alternatives

Instead of creating a layered approach, we could have simply created a flat`transaction message`, that is tailored for mutating the ledger state, and try to fit all the use cases there. For example with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.

This approach seems less extensible. It might have made sense if we wanted to build a protocol that is just for ledger mutating transactions, but we want to be able to extend the protocol to do more than that.

# Unresolved questions

- What should be the maximum length of the message?
- Better protection with PoW? how does the length of the message and type of payloads should affect difficulty?
- How should we pad the message to fit it into CURL for PoW?

> - How should we pad the input for CURL? One proposal is to go with padding method 2 that is described here https://en.wikipedia.org/wiki/ISO/IEC_9797-1#Padding

Author: I was trying to get some response from researchers about this one but maybe I didn't do it hard enough

@Wollac offred the padding rule and I think it is good enough because even if someone manages to speed up PoW it won't be the most troubling thing

Reviewer 5: Has been resolved with #24

> - How should we pad the message to fit it into CURL for PoW?
> - How should we pad the input for CURL? One proposal is to go with padding method 2 that is described here https://en.wikipedia.org/wiki/ISO/IEC_9797-1#Padding
> - Do we want to move away from using CURL for PoW in the future?

Reviewer 5: ```suggestion
```
Should be removed as PoW is now discussed in #24

| ByteArray    | A dynamic size byte array. It is a prefixed by a `varint` that indicates the number of bytes in the array. The following bytes are simply the content of the array.
| ByteArray[N] | A static size array of size N. Since the size is expected, no need to prefix with a varint   |

### Message ID
The message ID will be the `BLAKE2b-256` hash of the byte contents of the message. It should be used by the nodes to index the messages and by external APIs.

### Message Structure

<table>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Version</td>
        <td>varint</td>
        <td>The message version. The schema specified in this RFC is for version <strong>1</strong> only. </td>
    </tr>
    <tr>
        <td>Parent1 (<code>trunk</code>)</td>
        <td>ByteArray[32]</td>
        <td>The Message ID of the first <i>Message</i> we reference.</td>
    </tr>
    <tr>
        <td>Parent2 (<code>branch</code>)</td>
        <td>ByteArray[32]</td>
        <td>The Message ID of the second <i>Message</i> we reference.</td>
    </tr>
    <tr>
        <td>Payload Length</td>

>         <td>varint</td>

Reviewer 7: What to we gain by using a varint here? Are you planning to have megabyte-sizes messages? Make it a uint32 and reduce complexity instead of making even the header of the message variable in size

Author: We wanted to not have too many types in the message.
The idea is that less types means more simple descriptions and code.

Also, for lots of cases where we describe quantities `uint8` or `uint16` could have sufficed.

So the idea is that choosing varints will result mostly in shorter messages

Reviewer 9: I'm not convince it will be simpler code. It's not native primitives in many languages.
So we either implement one ourselves or use the existing one. Both would end up with more validation rules.

Reviewer 3: If `uint8_t` or `uint16_t` is sufficed, why not just use one of them? they are primitive types in most languages. The `varint` is good for data serialization but it is not easy to operate and memory management in a data structure, especially for microcontrollers.

Author: due to recent conversations outside of github I am also leaning towards dropping varints

this decision will be finalized on the next protocol team meeting

Reviewer 10: TL;DR: I would rather set fixed reasonable-sized types with some slack (you never know) replacing all varints.

`varints` are somewhat complex types to parse, but I would like to draw your attention to the creation of such message rather than the parsing of it. `varints` type force the entire message structure to "shift" instead of having fixed offsets. This is quite cumbersome when forming a message where all the parts are not necessarily known before-hand, and therefore the byte-requirements of length `varint` types is also unknown: when you fill in the data section of the message, you may see yourself push it forward in order to accommodate the corresponding `varint` size without overflowing into the data section itself. Sure, modern programming languages and libs provide abstract structures for all these, but at a cost of memory reallocation, memory concatenation and other forms of memory juggling. Leveling the participation field ultimately participates in the flexibility of a solution.

                        <th>Description</th>
                    </tr>
                    <tr>
                        <td>Payload Type</td>
                        <td>varint</td>
                        <td>
                            The type of the payload. It will instruct the node how to parse the fields that follow. Types in the range of 0-127 are "core types" that all nodes are expected to know.
                        </td>
                    </tr>
                    <tr>
                        <td>Data Fields</td>
                        <td>ANY</td>
                        <td>A sequence of fields, where the structure depends on <code>payload type</code>.</td>
                    </tr>
                    <tr>
                        <td>Nested Payload (Optional)</td>
                        <td> Payload</td>
                        <td> Some payloads support nesting other payloads inside of them.</td>
                    </tr>
                </table>
            </details>
    <tr>
        <td>Nonce</td>
        <td>uint64</td>
        <td>The nonce which lets this message fulfill the Proof-of-Work requirement.</td>
    </tr>
</table>

> ### Message PoW Hash

Reviewer 5: We need to explain this in more detail:
- We must have at least one example for the bytes of a serialized message.
- We need to carefully describe:
  - how the padding works
  - whether that is weighted by the length or something
  - how the difficulty is computed from the Curl hash
- Provide examples of PoWed data of different difficulty

### Message Validation

A message is considered valid, if the following syntactic rules are met:

1. The message length must not exceed X [tbd] bytes.
2. Varint fields must adhere to the following rules:
    * *Malleability protection:* No `0x80` or `0x00` trailing bytes. Take for example the following 3 byte varint: `0x8F8000`. 
    In order to parse it we look at its bits: 
    `1000 1111 1000 0000 0000 0000`.
    First we strip the msbs:
    `000 1111 000 0000 000 0000`
    Then we reverse the groups of 7 bits (Little Endian):
    `000 0000 000 0000 000 1111`
    Then we concatenate all the bits to finally get:
    `1111`
    So the varint `0x0F` equals to `0x8F8000`, but is encoded with different bytes.
    * *Size Protection*: The varint must not be more than 10 bytes long. A 10 byte varint can encompass all the range of `uint64`. In case of an overflow above the `uint64` range the message is invalid.
2. When we are done parsing the message there shouldn't be any trailing bytes left that were not parsed.

> 4. If the `payload type` is in the core payload range (0-127) and the node is familiar with it, or if it is in the range of 128-16383. The upper limit of 16383 is the maximal value of a 2 byte varint, and it should suffice for all the optional payload types we need.

Reviewer 5: Limiting the number to 16383 or 2 bytes seems rather arbitrary. We have a varint anyways so way not use its power to potentially store much higher number when needed.

Reviewer 11: I agree. ☝️ I see this artificial limitation as potentially shooting ourselves in the foot like with what happened during y2k.

> 5. If the payload is valid according to the **syntactic** rules the `payload type` defines.

Reviewer 7: This actually breaks the layer concept. A message should be valid it **its** structure is valid, regardless of the payload.

Author: Yup

This can actually be a problem the way it is defined here. If nodes will validate payloads differently (i.e. a newer node can know of a new payload type), this will affect the message...

I will remove this line
